### ISSUE TITLE
ncclient cannot server multiple venders within same process

### ISSUE THREAD

**Issue by singlaive** [ncclient/ncclient#386]
My application manages switches from different vendors. It is a software service with multithreading architecture, with each thread managing its own ncclient manager instance, and access to switches from different venders. What happens is once one thread finishes one job for vender A, the other thread starts another job for vender B, it will fail.
The reason is quite obvious, the VENDER_OPERATION defined at https://github.com/ncclient/ncclient/blob/master/ncclient/manager.py#L58 is a global, and would comes with different methods per vender.

I got a feeling this is designed so. Can anyone suggest or confirm so? Or any way to get around it?

**Comment by psikala** [comment]
How about sharing some code how to reproduce the problem?

**Comment by psikala** [comment]
How about sharing some code how to reproduce the problem?

**Comment by psikala** [comment]
Here is a quick 5 minute example, which get's configuration from Huawei, Alcatel and Juniper -devices using multiprocessing Pool, ProcessPoolExecutor & ThreadPoolExecutor. Everything works as expected. When using processes, result should be returned for ex. as a string (`result.data_xml`), because lxml.etree._Element -objects cannot be pickled.

```python
import logging
from multiprocessing import Pool
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor

from ncclient import manager
from ncclient.xml_ import *

POOL_SIZE = 5


class NetconfDevice:

    def __init__(self, ip, port, driver, username, password):
        self.ip = ip
        self.port = port
        self.driver = driver
        self.username = username
        self.password = password


def get_configuration(device: NetconfDevice):
    conn = manager.connect(
        host=device.ip,
        port=device.port,
        username=device.username,
        password=device.password,
        hostkey_verify=False,
        device_params={'name': device.driver},
        timeout=240,
        ssh_config='~/.ssh/config'
    )
    result = False

    if device.driver == 'huaweiyang':
        FILTER = '''<ifm xmlns="urn:huawei:yang:huawei-ifm">
                    <interfaces>
                        <interface>
                            <name>NULL0</name>
                        </interface>
                    </interfaces>
                </ifm>'''

        result = conn.get(('subtree', FILTER)).data_xml

    elif device.driver == 'alu':
        FILTER = '''<configure xmlns="urn:alcatel-lucent.com:sros:ns:yang:conf-r13">
                 <system/>
            </configure>'''

        result = conn.get_configuration(filter=FILTER).data_xml

    elif device.driver == 'junos':
        FILTER = '''<configuration>
                <interfaces/>
            </configuration>'''
        result = conn.get_configuration(filter=to_ele(FILTER)).data_xml

    conn.close_session()
    return result


if __name__ == '__main__':
    log_format = '%(asctime)s,%(msecs)-3d %(levelname)-8s %(threadName)s PID-%(process)d : %(message)s'
    logging.basicConfig(stream=sys.stdout, level=logging.INFO, format=log_format)

    device_list = [
        NetconfDevice('192.168.1.1', 830, 'huaweiyang', 'username', 'password'),
        NetconfDevice('192.168.1.2', 830, 'alu', 'username', 'password'),
        NetconfDevice('192.168.1.3', 22, 'junos', 'username', 'password')
    ]

    with Pool(processes=POOL_SIZE) as pool:
        pool.map(get_configuration, device_list)

    with ProcessPoolExecutor(max_workers=POOL_SIZE) as pool:
        pool.map(get_configuration, device_list)

    with ThreadPoolExecutor(max_workers=POOL_SIZE) as pool:
        pool.map(get_configuration, device_list)

```

**Comment by singlaive** [comment]
Thanks for you reply. I have no doubt your snippet would work. Because it multi processes, and obviously, they do not share the VENDER_OPERATION global variable. But I am talking about multi threading, and all threads access to the same variable obviously. For ncclient extensions, like Huawei, Juniper etc., when each instance of device comes to line https://github.com/ncclient/ncclient/blob/master/ncclient/manager.py#L260, the code will visit the content of VENDER_OPERATION lastly assigned.

I will be very happy to hear I am horribly wrong here.

**Comment by psikala** [comment]
Now I got it.

That's true. There is a risk that operations with same fingerprint become overwritten. In that sense, ncclient is not thread -safe. Manager needs some refactoring.

**Comment by singlaive** [comment]
Do you know how issues get proceeded in ncclient project? Any decision maker needed to be @ here to for further discussion?

**Comment by einarnn** [comment]
@singlaive, while not one of the original authors of the library, I'm currently the primary maintainer of ncclient. I took a look at the code you referred to, and, yes, I can see the problems. This isn't, I think you'll agree, a thread safety issue, or anything to do with threading specifically. It is a product of the `VENDOR_OPERATIONS` dictionary being global as opposed to being connection instance data when the `connect` function is called.

If you wish to propose some refactoring via a pull request I will be very happy to review it. Any change would need to be backwards compatible with respect to the external interfaces of ncclient.

**Comment by singlaive** [comment]
@einarnn, thanks for you confirmation. Yes my company behind me will sponsor the fix to this bug. What is the typical time frame from PR accepted to new version of ncclient officially released?

**Comment by einarnn** [comment]
The last release was at the end of December, and we have had a number of PRs since then, but I don't have an official release cadence. If this issue was addressed, I'd probably make a release soon after the PR was accepted.

**Comment by andrewmallory** [comment]
I've run into this too in our multi-vendor non threaded operations that reuse existing python processes to manage multiple vendors using celery. I submitted pull request 392 with a fix of simply moving the global variable into an instance specific one and tested where we have had failures.

**Comment by einarnn** [comment]
Have now merged #392, so closing this issue.
